baseband/vlbi_base/base.py

```diff
@@ -512,6 +500,55 @@ def _unsqueeze(self, data):
                 new_shape.insert(i + 1, 1)
         return data.reshape(new_shape)
 
+    def write(self, data, invalid_data=False):
+        """Write data, buffering by frames as needed.
+
+        Parameters
+        ----------
+        data : array
+            Piece of data to be written, with sample dimensions as given by
+            ``sample_shape``. This should be properly scaled to make best use
+            of the dynamic range delivered by the encoding.
+        invalid_data : bool, optional
+            Whether the current data is invalid.  Defaults to `False`.
+        """
+        assert data.shape[1:] == self.sample_shape, (
+            "'data' should have trailing shape {}".format(self.sample_shape))
+
+        if self.squeeze:
+            data = self._unsqueeze(data)
+
+        count = data.shape[0]
+        offset0 = self.offset
+        sample = 0
+        while count > 0:
+            frame_index, sample_offset = divmod(self.offset,
+                                                self.samples_per_frame)
+            if frame_index != self._frame_index:
+                frame = self._frame = self._make_frame(frame_index)
```

Comment:
Not sure how happy I am with having `frame`, `self._frame` and possibly also `self._data`, or the fact that when writing a frame we do, say `assert frame is self._frame`.